File: job-search/serivces/Job-recommender/testing_with_python/TF-IDFCalculator.py
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity
import re
from nltk.corpus import stopwords
from nltk.stem import PorterStemmer

# Your data
with open('JobTitleEng.txt', 'r', encoding='utf-8') as file:
    job_titles = [line.strip() for line in file]

# Initialize the stemmer
stemmer = PorterStemmer()

# Define a function to preprocess the text
def preprocess(text):
    # Lowercase the text
    text = text.lower()
    # Remove special characters
    text = re.sub(r'\W', ' ', text)
    # Remove single characters
    text = re.sub(r'\s+[a-zA-Z]\s+', ' ', text)
    # Remove single characters from the start
    text = re.sub(r'\^[a-zA-Z]\s+', ' ', text) 
    # Substitute multiple spaces with single space
    text = re.sub(r'\s+', ' ', text, flags=re.I)
    # Remove prefixed 'b'
    text = re.sub(r'^b\s+', '', text)
    # **Remove numbers**
    text = re.sub(r'\d+', '', text)
    # Remove stopwords and stem the words
    return text

# Preprocess the job titles
job_titles = [preprocess(title) for title in job_titles]
# Create the Transform
vectorizer = TfidfVectorizer()

# Tokenize and build vocab
vectorizer.fit_transform(job_titles)
str1= "BackEnd Developer web"
str2 = " developer backend web java spring boot"
print(preprocess(str1))
print(preprocess(str2))

print(vectorizer.transform([preprocess(str1)]))
print("------------------------------------------------------------------------")
print(vectorizer.transform([preprocess(str2)]))

# Calculate Cosine Similarity
cos_sim = cosine_similarity(vectorizer.transform([preprocess(str1)]), vectorizer.transform([preprocess(str2)]))
print(cos_sim)

File: job-search/serivces/Job-recommender/nodejs_webapp_version/TF-IDFCalculator.js
const fs = require('fs');
const natural = require('natural');
const TfIdf = natural.TfIdf;
const tokenizer = new natural.WordTokenizer();
const tfidf = new TfIdf();

// Your data
let job_titles = fs.readFileSync('JobTitleEng.txt', 'utf8').split('\n');

// Preprocess the job titles
job_titles = job_titles.map(title => preprocess(title));

// Tokenize and build vocab
job_titles.forEach((title) => {
    tfidf.addDocument(tokenizer.tokenize(title));
});

let str1 = "backend developer";
let str2 = "backend developer nodejs typescript";

console.log(preprocess(str1));
console.log(preprocess(str2));

//console.log(tfidf.tfidfs(preprocess(str1)));
console.log("------------------------------------------------------------------------");
//console.log(tfidf.tfidfs(preprocess(str2)));
tfidf.tfidfs(preprocess('developer'), function(i, measure) {
    console.log('document #' + i + ' is ' + measure);
});
// showTfidfValues(tfidf, preprocess(str2));// Calculate Cosine Similarity
// let cos_sim = cosineSimilarity(tfidf.tfidfs(preprocess(str1)), tfidf.tfidfs(preprocess(str2)));
// console.log(cos_sim);

function preprocess(text) {
    // Lowercase the text
    text = text.toLowerCase();
    // Remove special characters
    text = text.replace(/\W/g, ' ');
    // Remove single characters
    text = text.replace(/\s+[a-zA-Z]\s+/g, ' ');
    // Remove single characters from the start
    text = text.replace(/^\^[a-zA-Z]\s+/g, ' '); 
    // Substitute multiple spaces with single space
    text = text.replace(/\s+/g, ' ');
    // Remove numbers
    text = text.replace(/\d+/g, '');
    return text;
}

function cosineSimilarity(vecA, vecB) {
    console.log("xnxx:", vecA)
    let dotProduct = 0;
    let normA = 0;
    let normB = 0;
    for (let i = 0; i < vecA.length; i++) {
        dotProduct += vecA[i] * vecB[i];
        normA += Math.pow(vecA[i], 2);
        normB += Math.pow(vecB[i], 2);
    }
    return dotProduct / (Math.sqrt(normA) * Math.sqrt(normB));
}

function showTfidfValues(tfidf, text) {
    console.log("xxxxxxxxxxxxxxxxxxxx")
    let tokenizedText = tokenizer.tokenize(text);
    tfidf.listTerms(tfidf.documents.length - 1).forEach(function(item) {
        if (tokenizedText.includes(item.term)) {
            console.log(item.term + ': ' + item.tfidf);
        }
    });
}